src/FsAutoComplete.Core/State.fs

```diff
@@ -38,6 +41,16 @@ type State =
     x.Files.[file] <- fileState
     x.FileCheckOptions.[file] <- opts
 
+  member x.AddCancellationToken(file : SourceFilePath, token: CancellationTokenSource) =
+    let lst = x.CancellationTokens.GetOrAdd(file, fun _ -> [])
+    let lst' = token::lst
+    x.CancellationTokens.[file] <- lst'
+
+  member x.ClearCancellationTokens(file : SourceFilePath) =
+    x.CancellationTokens.[file] <- []
```

Comment:
I'd suggest to remove that or make it atomic in combination with "Cancellation" so that we cannot remove "uncancelled" tokens by accident